### ISSUE TITLE
bcftools mpileup

### ISSUE THREAD

**Issue by suswalsh** [samtools/bcftools#852]
I am running through this pipeline instead of using samtools and the diff in DP is huge, this must be due to bcftools filtering of everything! The use and explanation of --ff and --rf is not helpful, can these be fixed? I really would prefer to filter after with my own choices than have bcftools version of mpileup do it! Cheers for the help!

**Comment by pd3** [comment]
I don't really understand the question, can you be more specific?

**Comment by suswalsh** [comment]
Sorry I should be more clear, I was using samtools mpileup in my pipeline for calling variants (and normal bases) in amplicon sequencing. It worked well but due to the warning that is now popping up about it being deprecated, I wanted to move to bcftools sooner than later. My code for samtools was 
samtools mpileup -uf ref.fasta  -g sorted.bam | bcftools call -m -M > sorted.vcf

I hope I'm not making a glaringly obvious error, but my code for bcftools is 
bcftools mpileup -f ref.fasta sorted.bam | bcftools call -M -m > sorted.vcf 

The same file goes from DP 1440 (samtools) to DP 43 (bcftools), something I had thought to do with perhaps removal of duplicates or some sort of filtering in my change over to bcftools, unless of course I am doing something else incorrect with the code?
I have tried looking at -Q, --rf --ff, -C, -A, -B, and nothing but a few minor changes, by a couple of reads in DP changes. Im quite unsure where I am going wrong here and any help would be GREATLY appreciated. I understand that both algorithms for mpileup would give slightly different results but not to that extreme! Thanks for reaching out to help! Cheers!

**Comment by suswalsh** [comment]
oh and forgot to mention that DP was for a singular snp i was interested in.

**Comment by pd3** [comment]
The `samtools mpileup` and `bcftools mpileup` should give about the same result. You can debug the problem by leaving out the `bcftools call` command for now and check the difference in the outputs. Best if you post the two lines including any output from the commands.

**Comment by suswalsh** [comment]
I ran both and really the only diff I can see is the direction of the Alt - ref calls.  ( I ran bcftools view with samtools mpileup to see the output - this was not needed for bcftools mpileup as i could read the output without view).
It seems in bcftools it is not calling all of the sites, samtools is calling all alt alleles. see below.

For the same bam file at a singular site example.
bcftools mpileup
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	sample
15:28120272-28120672	201	.	A	G,<*>	0	.	DP=43;I16=7,1,33,1,257,8917,1339,54523,480,28800,2040,122400,200,5000,850,21250;QS=0.162864,0.837136,0;VDB=8.91952e-17;SGB=-0.693132;RPB=0.64958;MQB=1;MQSB=1;BQB=0.0649748;MQ0F=0	PL	216,0,81,240,183,255

samtools mpileup
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	sample
15:28120272-28120672	201	.	A	G,C,<*>	0	.	DP=1440;I16=267,2,875,4,11901,617089,41885,2.25394e+06,16140,968400,52740,3.1644e+06,6725,168125,21975,549375;QS=0.223899,0.767826,0.00827522,0;VDB=0;SGB=-0.693147;RPB=0.99932;MQB=1;MQSB=1;BQB=1.15554e-07;MQ0F=0	PL	255,0,158,255,255,255,255,255,255,255

can anyone assist with fixing this so the DPs are more alike? Thanks!

**Comment by suswalsh** [comment]
I am still having the above problems with extreme differences in DP when comparing samtools to bcftools, can anyone help with this problem? Thanks!!

**Comment by pd3** [comment]
Can you share the slice of BAM at this position with the full header? My best guess is that this could be a problem with different handling of readgroups and samples.

**Comment by suswalsh** [comment]
Well its the same ID in the 'sample' column. my BAM file only has a singular sample and I added readgroup ids to the bam in a prior set with picard tools. At no point do I have multiple samples in the BAM. Also the sample column after FORMAT is not the name of the file as that is has sorted.bam in it, it is the readgroup name i added for both. Im just not sure why it is different.

**Comment by suswalsh** [comment]
@RG	ID:foo	LB:bar	PL:illumina	SM:9948-1n-1	PU:unit1

**Comment by suswalsh** [comment]
even if i remove add RG step, still the same issue with the DP difference.

**Comment by pd3** [comment]
We'd need a reproducible test case in order to be able to help with this.

**Comment by suswalsh** [comment]
It is happening with all the bam files I have, which are simply aligned to ucsc hg19. Im sure you will generate the same DP difference with any bam you have, however if you need me to provide one, how would you like me to do this?

**Comment by pd3** [comment]
Best if you could run `samtools view -b -o test.bam problematic.bam 15:28120272-28120672` and make the bam available. If it is sensitive data, you could encrypt it and send it offline. My email is on my profile page. Thank you

**Comment by suswalsh** [comment]
I dont understand what you mean by test.bam v problematic.bam, it is only one bam once using samtools mpileup gives a similar number of DP when viewed in IGV and then bcftools mpileup when run on the same bam gives one tenth the number of reads when compared to IGV. I will send you the bam now, thanks.

**Comment by pd3** [comment]
Thank you for the testing bam. It turns out the difference is caused by `bcftools mpileup -d` behaving differently from `samtools mpileup -d`.

Previously the command had a minimum value of `8000/n` where `n` was the number of input files. This means that the default was increased from 250 to 8000 in your case. (The `-d` parameter would have an effect only once above the cross-sample minimum.) This behavior is problematic when working with a combination of single- and multi-sample bams and impossible to override, therefore it was changed in bcftools.

So, if you run as `bcftools mpileup -d 8000`, you should get the same result. 

I just updated the man page to explain this and added a message to notify about this default limit: 
`[mpileup] maximum number of reads per input file set to -d 250`. So hopefully no one else will get bitten by this again!

If you are wondering why the DP in your output was even lower than 250, that's because the decision about adding a read to the pileup is made only once, at the position where the read starts. If there are many reads piled up at one position, the following positions may be underrepresented.

**Comment by suswalsh** [comment]
fantastic Petr, thanks so much for your help with this!

**Comment by ALFLAG** [comment]
Hi Petr, @pd3 
I 'm afraid there's still an issue with the -d option.

Here the thing.
In one genome position (26464), most of the reads have a deletion of one nucleotide, while a tiny minority of the reads have the nucleotide that is present in the reference.

When I use bcftools mpileup with setting the -d option to 8000 (and higher), the deletion is not found. Instead, the command outputs the minority base as the most present.
When I remove the -d option from the command line, so the default value is set, the problem dissappeared. The deletion was well found
When I set the -d option to 250, the deletion was well found.

leaving the -d option to the default value is not suitable for me, so do you have an explanation for this behavior?

bcftools 1.9
samtools 1.9

Here the complete command I used
```bcftools mpileup -f ref.fasta -Q 22.5 -q 5 -d 8000 --threads 12 -O z -o pileup_file.gz sorted_bamfile.bam```

and input files
[sorted_alignment_on_viral_ref.bam.gz](https://github.com/samtools/bcftools/files/4293996/sorted_alignment_on_viral_ref.bam.gz)
[BestRefMegaB.fasta.gz](https://github.com/samtools/bcftools/files/4293998/BestRefMegaB.fasta.gz)

**Comment by pd3** [comment]
Please try if your problem is present with the latest github version. If it does, please do not add to a closed thread but open a new issue instead.
